.circleci/config.yml

```diff
@@ -93,53 +117,187 @@ commands:
+      - restore_cargo_package_cache
+      - run:
+          name: Build release
+          command: RUST_BACKTRACE=1 cargo build -j 16 --release
+      - build_teardown
+  build-e2e-test:
+    executor: build-executor
+    description: Generate a list of E2E test targets to be distributed.
+    steps:
+      - build_setup
+      - attach_workspace:
+          at: /home/circleci/project
+      - restore_cargo_package_cache
+      - run:
+          name: Find all e2e tests
           command: |
-            [[ $CIRCLE_NODE_INDEX =~ [0234] ]] || RUST_BACKTRACE=1 cargo build -j 16 --release
+            cargo x test --package testsuite -- --list | grep "::" | \
+              sed 's/: .*$//' > e2e_tests
+            cat e2e_tests
+      - persist_to_workspace:
+          root: /home/circleci/project
+          paths:
+            - e2e_tests
+            - target/debug/deps/libratest*
+  run-e2e-test:
+    executor: build-executor
+    parallelism: 8
+    description: Run E2E tests in parallel. Each container runs a subset of
+      test targets.
+    steps:
+      - build_setup
+      - attach_workspace:
+          at: /home/circleci/project
       - run:
-          name: Build Dev
+          name: Determine test targets for this container.
+          # NOTE Currently the tests are distributed by name order. Once test
+          # metadata is enabled, the tests can be distributed by run time to
+          # speed up this job.
           command: |
-            [[ $CIRCLE_NODE_INDEX =~ [0124] ]] || RUST_BACKTRACE=1 cargo build -j 16
-            [[ $CIRCLE_NODE_INDEX =~ [0124] ]] || RUST_BACKTRACE=1 cargo build -j 16 -p libra-swarm
-            [[ $CIRCLE_NODE_INDEX =~ [0124] ]] || RUST_BACKTRACE=1 cargo build -j 16 -p cluster-test
-            [[ $CIRCLE_NODE_INDEX =~ [0124] ]] || RUST_BACKTRACE=1 cargo build -j 16 -p libra-fuzzer
-            [[ $CIRCLE_NODE_INDEX =~ [0124] ]] || RUST_BACKTRACE=1 cargo build -j 16 -p language_benchmarks
-            [[ $CIRCLE_NODE_INDEX =~ [0124] ]] || RUST_BACKTRACE=1 cargo build -j 16 -p cost-synthesis
-            [[ $CIRCLE_NODE_INDEX =~ [0124] ]] || RUST_BACKTRACE=1 cargo build -j 16 -p test-generation
+            ls -all target/debug/*
+            echo -e "Found $(wc -l e2e_tests) test."
+            cat e2e_tests | circleci tests split > /tmp/tests_to_run
+            echo -e "This runner will run these tests\n$(cat /tmp/tests_to_run)"
       - run:
-          name: Run All Non Flaky Unit Tests
+          name: Run E2E tests
           command: |
-            [[ $CIRCLE_NODE_INDEX =~ [0134] ]] || RUST_BACKTRACE=1 $CI_TIMEOUT cargo test --all-features --workspace --exclude libra-node --exclude libra-crypto --exclude testsuite --exclude consensus
+            set -x
+            export libratest=$(find target/debug/deps/ -name "libratest*" \
+              -executable)
+            for target in $(cat /tmp/tests_to_run) ; do
+              if [ ! -z "$libratest" ]; then
```

Comment:
Why wouldn't libratest be there? It seems weird to build it and then continue on if the previous step somehow failed.

Reply:
Glad you called this out.  This is to account for infra failure on circle's side when the saving/attaching workspace fails silently in the backend.  Our build/test will complete in spite of that.  

On the flip side, we won't be able to detect such infra failure on circle's side, unless we put in telemetry to measure each job's metrics.

Reply:
Looking at it 2nd time, it looks like I can post these infra failures through slack webhook. I will follow this up in another commit.